gsuite_reports_rules/gsuite_drive_visibility_change.py

```diff
@@ -1,42 +1,123 @@
+
+            change_document_visibility = True
+            break
+
+    # "change_document_access_scope" events are always paired with "change_document_visibility" events
+    # the "target_domain" and "visibility" attributes are the same
+    # the set of a "change_document_access_scope" and a "change_document_visibility" is paired with a
+    # second set of these events
+    # the first pair of events represents the "old" domain and visibility, and the second pair
+    # represents the "new" domain and visibility
+    if change_document_visibility:
+        for details in event.get("events", [{}]):
+            if (
+                details.get("type") == "acl_change"
+                and details.get("name") == "change_document_access_scope"
+                and param_lookup(details.get("parameters", {}), "new_value") != ["none"]
+            ):
+                global ACCESS_SCOPE
+                ACCESS_SCOPE = param_lookup(details.get("parameters", {}), "new_value")
+        return True
+
+    # TODO: confirm that change_user_access events are mutually exclusive with the above events
+    # split this out
+    #########
+    # for target_user
+    change_user_access = False
+
     for details in event.get("events", [{}]):
         if (
             details.get("type") == "acl_change"
-            and param_lookup(details.get("parameters", {}), "visibility_change") == "external"
+            and details.get("name") == "change_user_access"
+            and param_lookup(details.get("parameters", {}), "new_value") != ["none"]
         ):
-            return True
+            global TARGET_USER_EMAIL
+            TARGET_USER_EMAIL = param_lookup(details.get("parameters", {}), "target_user")
+            DOC_TITLE = param_lookup(details.get("parameters", {}), "doc_title")
+            ACCESS_SCOPE = param_lookup(details.get("parameters", {}), "new_value")
+
+            change_user_access = True
+            break
+
+    if change_user_access:
+        return True
 
     return False
 
 
 def dedup(event):
-    doc_title = "<UNKNOWN_DOC_TITLE>"
-    for detail in event.get("events", [{}]):
-        if detail.get("type") == "acl_change":
-            if param_lookup(detail.get("parameters", {}), "doc_title"):
-                doc_title = param_lookup(detail.get("parameters", {}), "doc_title")
-                break
-    return doc_title
+    return DOC_TITLE
 
 
 def title(event):
-    target_user_email = "<EMAIL_UNKNOWN>"
-    doc_title = "<UNKNOWN_DOC_TITLE>"
-    for detail in event.get("events", [{}]):
-        if detail.get("type") == "acl_change":
-            if param_lookup(detail.get("parameters", {}), "doc_title"):
-                doc_title = param_lookup(detail.get("parameters", {}), "doc_title")
-            if param_lookup(detail.get("parameters", {}), "target_user"):
-                target_user_email = param_lookup(detail.get("parameters", {}), "target_user")
-            break
+
+    if TARGET_USER_EMAIL != "<UNKNOWN_USER>":
```

Comment:
So this is a case where the bad practice (which is sadly a standard for our detections) of global mutables may be coming to bite you. You're assuming that if the current event being evaluated didn't cause the rule to set the `TARGET_USER_EMAIL` global variable, it will have its default value.

Imagine this scenario:
- `event1` is processed by `rule`. This results in the `rule` function setting `TARGET_USER_EMAIL` to  have the value  `bob.bobertson@runpanther.io`
- `event1` is processed by `title`. It sees that `TARGET_USER_EMAIL` has the value `bob.bobertson@runpanther.io` and does the right thing.
- in the same lambda invocation, `event2` needs to be processed
- `event2` is processed by `rule`. This results in the `rule`  function not changing `TARGET_USER_EMAIL` at all.
- `event2` is processed by  `title`. This code appears to be expecting that `TARGET_USER_EMAIL` has the default value, but in fact `TARGET_USER_EMAIL` still has the value `bob.bobertson@runpanther.io` which causes this `title` to be screwed up and misleading

The basic crux of the issue is that if two events are processed in the same lambda invocation, the global variables in this function are _not_ reset between events. I think you could prove this to yourself by carefully crafting the unit tests, in fact I think you should do that to help make sure you understand the issue.

The simplest solution is to have `rule` reset all the default values immediately when it starts. I'm sure there are more advanced solutions we could think up, but since these global variables are a hack regardless I wouldn't spend too much time thinking about it.

Reply:
Alternatively, would transitioning this to use `gsuite.activityevent` make doing this lookups trivial/remove the need to use globals?

Reply:
> in the same lambda invocation, event2 needs to be processed

Do you mean the next `report` (consisting of multiple `event`s)? I didn't know a single lambda environment persisted from one invocation of a rule to the next. Does that only happen for invocations of the same rule? If it can happen for different, completely random rules....that would seem like a good reason to never use globals. It's a potential information exfiltration route. One could imagine a scenario where an attacker identifies/guesses a global variable name, and then constructs other rules to push those global values to malicious alert destinations (or compromised internal destinations).

If you're actually talking about the `event`s inside a single `report`, I don't think additional `event`s would be processed after one triggers the rule.

Reply:
I added a `reset_alert_details` function that gets called at the beginning of the rule 

@glerb  you think this is ready to merge?

Reply:
@lindsey-w I ported over your strategy from standard_rules/unusual_login.py (which we discussed after this one). What do you think?

Reply:
@nhakmiller  is the scenario you describe in the top of this thread behaving as designed? It seems strange to mme that there is not more atomicity between each time a rule is called.